### PR TITLE
[Web UI] Network error improvements - offline and retry fields

### LINKED ISSUE #2423: [Web UI] Instrument http-link to toggle offline state in apollo cache

### PR DESCRIPTION
**Pull request by dabria**
This is >probably< to be merged in after GA

## What is this change?
Closes #2423 and #2422
Adds LocalNetwork to the Query component. LocalNetwork contains fields for offline and retry. These fields do nothing at this point in time and are an initial step towards #2103

## Why is this change necessary?
This is to help track the state of the network better and add appropriate errors for the user.

## Does your change need a Changelog entry?
I don't think so, when this functionality is completed in it's entirety then it would make more sense to add an entry. As of now these fields do nothing.

### CODE REVIEW

File: dashboard/src/apollo/client.js [hunk 1]
```diff
@@ -72,6 +72,29 @@ const createClient = () => {
...
+    client,
+    gql`
+      query SyncLocalNetworkQuery {
+        localNetwork @client {
+          offline
+          retry
+        }
+      }
+    `,
+  );
```
Comment (jamesdphillips) [inline dashboard/src/apollo/client.js]: Do these need to be persisted? Is this to synchronize the state across tabs?

File: dashboard/src/apollo/httpLink.js [hunk 2]
```diff
@@ -1,10 +1,34 @@
...
+const httpLink = ({ getClient }) =>
   new HttpLink({
     uri: "/graphql",
-    fetch: doFetch,
+    fetch: (url, init) =>
+      doFetch(url, init).then(
+        response => {
+          getClient().mutate({
+            mutation,
+            variables: { offline: true },
```
Comment (jamesdphillips) [inline dashboard/src/apollo/httpLink.js]: Unless I'm mistaking the intent, do we not want to set the state to false if the query was successful?

File: dashboard/src/apollo/schema/client.graphql [hunk 3]
```diff
@@ -29,6 +29,17 @@ extend type Query {
...
+  localNetwork: LocalNetwork
+}
+
+type LocalNetwork {
+  "The current online state of the client."
+  offline: Boolean!
+
+  "If the client is offline and trying to retry the connection."
+  retry: Boolean!
 }
```
Comment (jamesdphillips) [inline dashboard/src/apollo/schema/client.graphql]: 👍

File: dashboard/src/apollo/httpLink.js [hunk 4]
```diff
@@ -1,10 +1,34 @@
...
+          });
+          return response;
+        },
+        error => {
+          getClient().mutate({
+            mutation,
+            variables: { offline: false },
```
Comment (10xjs) [inline dashboard/src/apollo/httpLink.js]: James is right, these values are swapped.

File: dashboard/src/apollo/httpLink.js [hunk 5]
```diff
@@ -1,10 +1,34 @@
...
+      doFetch(url, init).then(
+        response => {
+          getClient().mutate({
+            mutation,
+            variables: { offline: true },
+          });
+          return response;
+        },
+        error => {
+          getClient().mutate({
```
Comment (10xjs) [inline dashboard/src/apollo/httpLink.js]: There is a check here that we are missing.

We need to test that the error is an instance of `FailedError` in order to isolate connection errors from other http errors.

(See https://github.com/sensu/sensu-go/blob/master/dashboard/src/utils/fetch.js for the heuristic)

File: dashboard/src/apollo/client.js (hunk 1 above)
Comment (10xjs) [reply dashboard/src/apollo/client.js]: I think this is unnecessary

File: dashboard/src/apollo/httpLink.js (hunk 2 above)
Comment (dabria) [reply dashboard/src/apollo/httpLink.js]: Whoops, yeah that should be false

File: dashboard/src/apollo/client.js (hunk 1 above)
Comment (dabria) [reply dashboard/src/apollo/client.js]: Was following an example previously, I've removed them

File: dashboard/src/apollo/httpLink.js (hunk 5 above)
Comment (dabria) [reply dashboard/src/apollo/httpLink.js]: Updated!

File: dashboard/src/apollo/resolvers/localNetwork.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  defaults: {
+    localNetwork: {
+      __typename: "LocalNetwork",
+      offline: false,
+      retry: false,
+    },
+  },
+  resolvers: {
+    Mutation: {
+      retryLocalNetwork: () => (_: mixed, args: mixed, { cache }: Context) => {
```
Comment (10xjs) [inline dashboard/src/apollo/resolvers/localNetwork.js]: There's an extra function wrapper here `() =>`

Comment (dabria) [reply dashboard/src/apollo/resolvers/localNetwork.js]: 🦅 👀

**Pull request merged**